### PR TITLE
feat: Added support for all proxy modes

### PR DESCRIPTION
**Pull request by LuoJinghua**
#### Description of Change

This pull request introduces following changes:
*  Extended the setProxy() to support all proxy modes including `direct`, `auto_detect`, `fixed_servers`, `pac_script` and `system`.
* Api for reloading proxy configuration.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).
- [x] This is **NOT A BREAKING CHANGE**. Breaking changes may not be merged to master until 11-x-y is branched.

#### Release Notes

Notes: Added support for explicitly specifying `direct`, `auto_detect` or `system` modes in `session.setProxy()`.

### CODE REVIEW

**Comment by deepak1556** [comment]
Thanks for the PR!

Can you split the `closeAllConnections()` and `closeIdleConnections()` to a separate PR.

Also can you add some tests, it would be difficult to test `system` and `auto_detect` mode but you can unittest the api and also verify the other modes.

File: shell/browser/api/electron_api_session.cc
Comment (deepak1556) [inline shell/browser/api/electron_api_session.cc:449]: "sysytem" => "system"

File: shell/browser/api/electron_api_session.cc
Comment (deepak1556) [inline shell/browser/api/electron_api_session.cc:444]: Can you use [StringToProxyMode](https://source.chromium.org/chromium/chromium/src/+/master:components/proxy_config/proxy_prefs.h;l=65) util and compare against the `ProxyPrefs::ProxyMode` , it will guarantee a level of upstream api breakages wrt the strings used.

Comment (LuoJinghua) [reply shell/browser/api/electron_api_session.cc:444]: ok.

**Comment by LuoJinghua** [comment]
> Thanks for the PR!
> 
> Can you split the `closeAllConnections()` and `closeIdleConnections()` to a separate PR.
> 
> Also can you add some tests, it would be difficult to test `system` and `auto_detect` mode but you can unittest the api and also verify the other modes.

Finished, please review my changes.

File: shell/browser/api/electron_api_session.cc
Comment (deepak1556) [inline shell/browser/api/electron_api_session.cc:491]: please use `auto* isolate = JavascriptEnvironment::GetIsolate();` instead

File: spec-main/api-session-spec.ts
Comment (deepak1556) [inline spec-main/api-session-spec.ts:449]: does this reload have any effect here ? I was imagining the reload is required after the config is updated.

```
const config = { mode: 'direct' as any, proxyRules: 'http=myproxy:80' };
await customSession.setProxy(config);
const proxy = await customSession.resolveProxy('http://example.com/');
expect(proxy).to.equal('DIRECT');
const newConfig = { mode: 'fixed_servers' as any, proxyRules: 'http=myproxy:80' };
await customSession.setProxy(config);
await customSession.forceReloadProxyConfig();
const newProxy = await customSession.resolveProxy('http://example.com/');
expect(newProxy).to.equal('PROXY myproxy:80');
```

Comment (LuoJinghua) [reply spec-main/api-session-spec.ts:449]: For this case it already works without forceReloadProxyConfig(). I think I just figured a ways to test this functionality.
```javascript
      let proxyPort = 8132;
      server = http.createServer((req, res) => {
        const pac = `
          function FindProxyForURL(url, host) {
            return "PROXY myproxy:${proxyPort}";
          }
        `;
        res.writeHead(200, {
          'Content-Type': 'application/x-ns-proxy-autoconfig'
        });
        res.end(pac);
      });
      await new Promise(resolve => server.listen(0, '127.0.0.1', resolve));
      const config = { mode: 'pac_script' as any, pacScript: `http://127.0.0.1:${(server.address() as AddressInfo).port}` };
      await customSession.setProxy(config);
      {
        const proxy = await customSession.resolveProxy('https://google.com');
        expect(proxy).to.equal(`PROXY myproxy:${proxyPort}`);
      }
      {
        proxyPort = 8133;
        customSession.forceReloadProxyConfig()
        const proxy = await customSession.resolveProxy('https://google.com');
        expect(proxy).to.equal(`PROXY myproxy:${proxyPort}`);
      }
```

Comment (deepak1556) [reply spec-main/api-session-spec.ts:449]: yeah that works too, as long as there is a change in configuration within the same session then testing this api makes sense.

File: shell/browser/api/electron_api_session.cc
Comment (LuoJinghua) [reply shell/browser/api/electron_api_session.cc:491]: Ok, got it.

File: spec-main/api-session-spec.ts
Comment (LuoJinghua) [reply spec-main/api-session-spec.ts]: > LGTM with minor nits

Fixed, could you please review my changes again.

**Comment by deepak1556** [comment]
changes look good, but this will need review from api working group before being merged.

**Comment by jkleinsc** [comment]
The @electron/wg-api approved this at the August 24, 2020 meeting

**Comment by LuoJinghua** [comment]
> The new APIs looks good to me, but I'd like to know a couple of things:
> 
> 1. Can the mode be set using the already-existing `proxyRules` string? Does the `mode` parameter provide access to functionality that's not currently accessible?


> 2. Can you say a little more about the use-case that motivates this API?
> 
> Thanks for the well-formatted PR (and test cases)!



> The new APIs looks good to me, but I'd like to know a couple of things:
> 
> 1. Can the mode be set using the already-existing `proxyRules` string? Does the `mode` parameter provide access to functionality that's not currently accessible?
If it's unspecified, it will be automatically determined based on other specified options and matches the current behaviour. The mode is set to pac_script if pacScript is specified, otherwise it's set to fixed_servers.

Without the mode parameter: 
* It's unclear how to set the proxy to direct mode, it's much clear to call `setProxy({ mode: 'direct' })` then `setProxy({proxyRules: ''})`
* There's no way to set proxy back to system/auto_detect mode after setting a proxy for a session

> 2. Can you say a little more about the use-case that motivates this API?
> 

Let me quote a part of reply from pull request #24945
> The real world use case is a user want to access to a site requires a proxy and he get a proxy list from somewhere.
> 1. Our product randomly select a proxy from the list and try to navigate the target site
> 2. Do some basic checks to verify whether the proxy is good and the website is accessable
> 3. Goto step 1 if the check failed
> 
> The problem is these proxies maybe from the same provider and the origin is same too. Everything is ok when the origin of desn't match previous one, otherwise the connection is reused when the server supports http 2 or keep-alived connections which are true for most of sites.

There's no way to switch the proxy back to system mode without this change to `setProxy` after finishing the dirty work.

> Thanks for the well-formatted PR (and test cases)!

File: docs/api/session.md
Comment (nornagon) [inline docs/api/session.md:195]:
```suggestion
      In fixed_servers mode the proxy configuration is specified in `proxyRules`.
      This is the default mode if `proxyRules` is specified.
```

File: docs/api/session.md
Comment (nornagon) [inline docs/api/session.md:193]:
```suggestion
      retrieved from the URL specified in the `pacScript`. This is the default mode
      if `pacScript` is specified.
```

File: docs/api/session.md
Comment (nornagon) [inline docs/api/session.md:283]: This needs a description of what the method does.

What configuration is reloaded? From where?

File: shell/browser/api/electron_api_session.cc
Comment (nornagon) [inline shell/browser/api/electron_api_session.cc:457]: Let's factor out the common part of these branches. e.g.

```cpp
std::unique_ptr<base::Value> proxy_config;
if (proxy_mode == ProxyPrefs::MODE_DIRECT) {
  proxy_config = ProxyConfigDictionary::CreateDirect();
} else if (...) {
  // ...
}
browser_context_->in_memory_pref_store()->SetValue(
    proxy_config::prefs::kProxy, proxy_config,
    WriteablePrefStore::DEFAULT_PREF_WRITE_FLAGS);
```

**Review by nornagon — CHANGES_REQUESTED** [review]
Thanks for the clarifications, and sorry for dropping the ball on this review. The general shape of the API looks good to me, just a couple of minor comments before we can get this landed.

**Review by nornagon — APPROVED** [review]

**Comment by nornagon** [comment]
@LuoJinghua looks like the build is failing, but otherwise lgtm :)

**Comment by LuoJinghua** [comment]
> @LuoJinghua looks like the build is failing, but otherwise lgtm :)

Fixed.

**Review by zcbenz — APPROVED** [review]

**Pull request merged**
